gel/protocol/codecs/object.pyx

```diff
@@ -348,23 +348,30 @@ cdef class ObjectCodec(BaseNamedRecordCodec):
             )
 
         lprops_type = None
-        proxy = getattr(return_type, '__proxy_of__', None)
-        if proxy is not None:
+
+        expr_object_types = getattr(return_type.__gel_reflection__, 'expr_object_types', None)
+
+        if proxy := getattr(return_type, '__proxy_of__', None):
             self.cached_return_type_proxy = return_type
             self.cached_return_type = proxy
             assert not hasattr(proxy, '__proxy_of__')
             lprops_type = return_type.__linkprops__
+            worklist = [self.cached_return_type]
+        elif expr_object_types is not None:
+            self.cached_return_type = return_type
+            self.cached_return_type_proxy = None
+            worklist = list(expr_object_types)
```

Comment:
Maybe name this something different than `worklist` here, and make a copy to populate the initial worklist?
(Instead of needing to recreate the `worklist` conditionally after the DFS loop)